More RSP doc additions

CQCumbers opened the pull request:
I've added descriptions of the MPEG instructions (VRNDP, VRNDN, VMULQ, and VMACQ), as well as fixed typos in the pseudo-code for some of the multiply instructions. The MPEG instruction behavior is based on Near's public domain code and PeterLemon's tests.

rasky commented:
Hello, thanks for the additions, I'll review them promptly.

It looks like your commits mistakenly revert this commit of mine: https://github.com/rasky/r64emu/commit/73ce551944534d7486afe7052d75e259eb215e8d#diff-994ec9a74052a8fc6042ed9a35b6248fde9f32479557208ef50f63b696ae80dd

Can you please fix it? Thanks

rasky commented on doc/rsp.md:
```diff
@@ -220,32 +220,42 @@ that the 32 registers are logically divided into 4 groups (0-7, 8-15, 16-23, 24-
+with `VPR[1]<0>`, to the following addresses: `$1E`, `$20`, `$22`, `$24`, `$26`, `$18`,
+`$1A`, `$1C`.
+
+`LTV` reads 128 bits from `(GPR[base] + (offset * 16)) & 0xFF8`, rotates them left by
+`element + (address & 0x8)` bytes, then loads the leftmost 16 bits into lane 0 of the
+specified diagonal, the next 16 bits into lane 1, and so on. If `element(0)` is 0
+and the address is 128-bit aligned, this has the effect of reading the given diagonal
+starting from register `v0` of the register group. For instance, `LTV v0[e2],$1E(r0)`
+reads diagonal 1, starting with `VPR[0]<7>`, from the following addresses: `$20`, `$22`,
+`$24`, `$26`, `$18`, `$1A`, `$1C`, `$1E`.
```
I don't specifically like this description because it is closer to how one would write an emulator for it, than a description of what the hardware does. I know it is a weird opcode, but I think we should strive to describe the hardware.

This is how I understand it:

```suggestion
`LTV` fetches two subsequent 64-bit aligned words, starting from `(GPR[base] + (offset * 16)) & 0xFF8`, and combine them into a 128-bit intermediate value called `W`, that will be used to populate the diagonal. The highest half in `W` is populated with the only 64-bit word among the two whose address is also 128-bit aligned. Lane 0 of the diagonal is then populated with the 16 bits in `W `starting at byte offset `element(3..0)` (`W[element(3..0)..element(3..0)+1]`). Subsequent lanes are populated with subsequent 16 bits words in `W`, wrapping around within `W`. For instance `LTV v0[e3],$1E(r0)` reads diagonal 1, starting with `VPR[0]<7>`, from `DMEM[$23..$24]`, `DMEM[$25..$26]`, `DMEM[$27,$18]`, `DMEM[$19..$1A]`, `DMEM[$1B..$1C]`, `DMEM[$1D..$1E]`, `DMEM[$1F..$20]`, `DMEM[$21..$22]`
```

(also notice I think your example was incorrect, I think I've fixed it but please double-check).

CQCumbers replied:
The suggestion is great and definitely less confusing. I think my example looks a little different because it starts with `VPR[0]<7>`, rather than `VPR[1]<0>`. In your suggestion did you meant to start with `VPR[1]<0>`? If so I think we are in agreement here.

rasky replied:
Uhm no, I would say that `VPR[0]<7> = DMEM[$23..$24]`. It looks like in your example (adjusted with `e3` instead of `e2`) you wrote that `VPR[0]<7> = DMEM[$21..$22]`. 

This is my understanding. Offset is $1E. The two 64-bit words are $18-1F and $20-$27. The second one is the 128-bit aligned one, so `W = DMEM[$20..$27,$18..$1F]`. `e3` correspond to an offset of 3 bytes, so the first 16-bit word is `W[3..4] = DMEM[$23..$24]`. `e3` also means diagonal 1; lane 0 of diagonal 1 is `<7>`, so the the first value `DMEM[$23..$24]` is written to `VPR[0]<7>`.

CQCumbers replied:
I used the notation `VPR[reg]<lane>`, so lane 0 of any diagonal is `VPR[diagonal]<0>`. As I understand it the value loaded into lane 0 stays the same for even `element` values, so that when you transpose a matrix the starting address you give is loaded into the appropriate lane of register 0, mirroring how the starting address you give `STV` is stored from lane 0 of the appropriate register.

rasky commented on doc/rsp.md:
```diff
@@ -254,21 +264,46 @@ the following sequence transposes it:
+`SWV` is much simpler than the other transpose instructions, and simply rotates `vt`
+left by `element` bytes before storing it to the address `GPR[base] + (offset * 16)`,
+wrapping at the second 64-bit boundary. The following sequence transposes a
+matrix stored at `$00(a0)..$7F(a0)`:
```
Please rephrase without "rotation", just describes which lanes go into which memory locations.

rasky commented on doc/rsp.md:
```diff
@@ -280,95 +315,168 @@ mapped into the 16-bit lanes. Signed opcodes (`LPV`, `SPV`) map the value to bit
-element access around (so basically the highest bit of `element` is ignored), while
-store instructions change the mapping bits, effectively swapping their meaning: `SPV`
-with `element` in the range `[8..15]` is like `SUV` with `element` in the range `[0..7]`,
-and viceversa.
```
I think I prefer the previous version. Can you please explain me what you found difficult to follow? I think your version is closer to how you would implement it, but mine is close to describing how the hardware behaves.

CQCumbers replied:
I think I read the original version as saying `SUV` would use the `(14..7)` mapping for all lanes, as long as the `element` value given in the instruction is in the range `[0..7]`. I wanted to make it clearer that when `element` is 1, the last lane will use the `(15..8)` mapping and the other lanes will use the `(14..7)` mapping.

I agree your load description is more clear. I tried using a "rotation" description because I thought some of the other instructions would be easier to describe after introducing the concept, but now that I see your `LTV` description I think it is possible to do without it.

rasky replied:
> I think I read the original version as saying SUV would use the (14..7) mapping for all lanes, as long as the element value given in the instruction is in the range [0..7].

You're right, this is not clear. Feel free to propose a new warning of that paragraph to clarify this.

CQCumbers replied:
Actually, I think the part that I found difficult to describe without rotation is packed loads with `element`, like `LPV v0[e1] $00(a0)`, where lane 0 is loaded from `$0e(a0)`.

rasky replied:
I think this should be explained by my sentence "`element` is used to select the first lane being accessed by the instruction (lane accesses also wrap around)".

Maybe we should further clarify it with "`element` is used to select the first lane with `vt` which is loaded with the first 16-bit word at the specified memory address. Subsequent 16-bit values are loaded into subsequent lanes in the register, wrapping around within the register".

CQCumbers replied:
Sorry, I misspoke. For packed loads specifically 16-bit lanes are loaded with consecutive 8-bit words, so `LPV v0[e1] $00(a0)` loads bytes `$0F,$00, $01,$02,$03,$04,$05,$06`. `$0F` is not subsequent to `$06`, and where this jump occurs depends on the load alignment. `LPV v0[e1] $01(a0)` is the same as `LPV v0[e0] $00(a0)`, for example.

rasky commented on doc/rsp.md:
```diff
@@ -280,95 +315,168 @@ mapped into the 16-bit lanes. Signed opcodes (`LPV`, `SPV`) map the value to bit
+                              then bytes $0b(a0),$0f(a0) into VPR[1]<4..5>
+    SFV v1[e5],$06(a0)     // write bytes $06(a0),$0a(a0),$0e(a0) from VPR[1]<7,4,5>
+                              then byte $02(a0) from VPR[1]<6>
 
```
I think this can go in as-is. I still don't specifically like the "rotation" approach to description, but since there was nothing before, I can't really complain, this is surely a good description anyway.

rasky commented on doc/rsp.md:
```diff
@@ -280,95 +315,168 @@ mapped into the 16-bit lanes. Signed opcodes (`LPV`, `SPV`) map the value to bit
+others. All 4 bits of `vt_elem` are used for this purpose. Note that the broadcast
+modifier `vt_elem` affects the value of `VT` whenever it is used within an instruction,
+but these modifications are not written back - that is, from the point of view of other
+instructions the value of `VT` remains the same.
```
Good catch! I had not realized that the `VMOV` access pattern was the same of the broadcast modifiers.

This said, are you sure this applies to *all* single-lane instructions? When I look at my code and my tests, I would say it only applies to `VMOV`. Can you provide a small snippet that shows this for any other single-large opcode?

CQCumbers replied:
Thank you so much for reviewing my additions, I really appreciate your insight.

If I understand you correctly, I don't think this contradicts your code or tests. Since lane `elem & 7` stays the same after applying broadcast modifier `elem`, I don't think it's possible to distinguish whether the modifier is applied or not. I suppose I thought it was simpler to describe it as applying to all the instructions.

rasky replied:
I'm still missing something, then. You're right that it's impossible to distinguish whether the broadcast modifiers is applied or not because `elem & 7` stays the same.

But then, this doesn't explain the hardware bug that I described here (in the part you're removed). That hardware bug mixes `vt_elem` and `vd_elem` together; it's true that the way they're mixed is somewhat related to the way broadcast modifier works, but it's still a behavior which is specific of `VMOV` and requires a special description, as far as I can tell.

Am I missing something?

rasky commented on doc/rsp.md:
```diff
@@ -280,95 +315,168 @@ mapped into the 16-bit lanes. Signed opcodes (`LPV`, `SPV`) map the value to bit
+    VD<de> = VT<de>
 
+As a side-effect, `ACCUM_LO` is loaded with `VT`. Note that the source and destination
+lanes are both `de`, and `vt_elem` is being used as a broadcast modifier.
```
Are you sure that `de` is also used for the source register after broadcast? Again this doesn't match my tests. Can you provide a RSP source code snippet that shows this?

CQCumbers replied:
I haven't run the test myself, but I based this description off of the screenshot in https://github.com/PeterLemon/N64/tree/master/RSPTest/CP2/LOADSTORE/TransposeMatrixVMOV. The VMOVs used to construct each row all have the same `se`, but different values are loaded into each lane. The pattern seems to match the behavior if the source lane is also `de`. (I am also assuming the screenshot is from real hardware)

rasky replied:
No, what you see in the upper half of that screenshot is the behavior of the hardware bug that I described in the text you removed. The code is here:

https://github.com/PeterLemon/N64/blob/63f09b404bef9c0bdba6be9efa87ce78401864d7/RSPTest/CP2/LOADSTORE/TransposeMatrixVMOV/RSPTransposeMatrixVMOV.asm#L948

Each VMOV has a fixed `se` per row, but the register changes so they're different values. `de` instead is < 8, so the hardware bug is triggered. This is why the bottom half of the screen (generated with `de` >= 8) shows a correct matrix transposition, while the upper half is wrong.

BTW I regenerated most of the screenshots in that repo using my own N64 and dumping the framebuffer content, so they're mostly pixel-perfect. My own emulator's testsuite compares the output pixel by pixel.

CQCumbers replied:
You can also explain the bottom half of the screen showing a correct matrix transposition with the broadcast modifier, since modifiers >= 8 broadcast one lane to the entire register. The first two rows of the top half are the same because modifiers 0 and 1 are the same. The entire test case  follows the pattern of applying each broadcast modifiers to the input matrix, then taking the diagonal.

I think your hardware bug replacing bits of `se` and the broadcast modifier might be different ways of describing the same thing, but am not 100% sure of that.

rasky replied:
Ah OK now I do understand. Your interpretation is actually made of two different parts: 1) vt_elem is used only as a broadcast modifier, 2) source and destination lane are the same.

This actually makes perfect sense and explains what I wrote as a "hardware bug". Kudos!

I think this should just be part of VMOV though. It's true that it's technically impossible to discern whether it applies to other single-instruction as you said, but at the same time the behavior makes sense only because there's also the part where source and dest lane are forced to be the same. I'd rather keep this whole weird behavior confined within VMOV.

Also I'd leave the broadcast table explanation later in the document. If we want, in a followup PR we can rearrange the order of the section so that it appears before single-lane instructions, but I wouldn't want to explain that in the context of VMOV. I'd rather it be part of the general description of computational instructions, and refer to it from VMOV.

rasky reviewed (commented):
Thanks! Lots of egregious work. My main doubt is about the wording you have been using that describes operations as "rotations". I've also provided a few suggestions on how to rephrase a few paragraphs. See below for more details.

CQCumbers commented:
Sorry for not communicating my progress as well as I should, this is my first time using Github's review system. I believe I've addressed your suggestions at this point (unless you've found the description of LTV or another instruction to be inaccurate). Thanks again for reviewing my changes, hopefully this can be merged soon.
